system/lib/libc/musl/src/thread/pthread_cond_timedwait.c

```diff
@@ -115,8 +115,10 @@ int __pthread_cond_timedwait(pthread_cond_t *restrict c, pthread_mutex_t *restri
 
 	__pthread_mutex_unlock(m);
 
+#ifndef __EMSCRIPTEN__
 	__pthread_setcancelstate(PTHREAD_CANCEL_MASKED, &cs);
 	if (cs == PTHREAD_CANCEL_DISABLE) __pthread_setcancelstate(cs, 0);
+#endif
```

Comment:
This looks like musl bug. According to POSIX spec https://pubs.opengroup.org/onlinepubs/9699919799/functions/V2_chap02.html "2.9.5 Thread Cancellation" pthread_cond_timedwait and pthread_cond_wait funcitons are cancellation points.

Test pthread_cond_wait/2-3 from Open POSIX suite checks such behavior and cancels thread waiting on conditional variable. With masked cancellation this test hangs.

Reply:
Hmm, if there is a musl bug, did you mark this as a `#ifndef __EMSCRIPTEN__` with the intent to "let the native side retain its buggy behavior to not break anything there" and fix the bug for Emscripten? Or was the divergence for some other reason?

This code does implement a `__pthread_testcancel()` call in the beginning of this file. It is not wrong for the implementation to mask cancellation afterwards. What kind of behavior are you seeing in the test suite with regards to this?

Reply:
Hmm, I wanted to mark Emscripten related changes/fixes with this macro. 

Problem is that test pthread_cond_wait/2-3 hangs, because there is scenario when thread waiting on condition variable is being cancelled. So if we set PTHREAD_CANCEL_MASKED for such thread it won't be cancelled.

I'll check behaviour of native musl with this test (pthread_cond_wait/2-3), maybe there is some other cancelation mechanism (like signals)...

Reply:
I've checked musl-1.2.0 behaviour and yes there is a bug there.

Original idea for this patch was to use `PTHREAD_CANCEL_MASKED` state distinguish:
- waits on functions which are non-cancellation points (e.g. `pthread_mutex_lock`) and can
  be canceled only when user wants asynchronous cancellation (i.e. sets thread
  cancelability type to `PTHREAD_CANCEL_ASYNCHRONOUS`). Such functions call `__timedwait()` to perform waiting.
- waits on functions which are defined as cancellation points (`pthread_cond_wait`).
  Such functions call `__timedwait_cp()`

This test was done by setting cancellation state to `PTHREAD_CANCEL_MASKED` and using  check:
```
    int can_cancel_wait = pthread_self()->canceldisable == PTHREAD_CANCEL_ENABLE ||
                         (pthread_self()->canceldisable == PTHREAD_CANCEL_MASKED &&
                          pthread_self()->cancelasync == PTHREAD_CANCEL_ASYNCHRONOUS);
```
in `__timedwait_cp`.

Original musl implementation which disables thread cancelability internally during call to waits (`__timedwait`), disables also user request to allow thread to be asynchronously canceled  (Cancelability Type set `PTHREAD_CANCEL_ASYNCHRONOUS`). This causes that test `pthread_setcanceltype/1-1` fails with native musl implementation.

As there are many functions (~30) in musl which in their implementation disables thread cancellation and additionally flag `PTHREAD_CANCEL_MASKED` was added to avoid some cancellation race in `pthread_cond_wait` in commit https://git.musl-libc.org/cgit/musl/commit/?id=8741ffe625363a553e8f509dc3ca7b071bdbab47, so in acd89cd I changed test for Emscripten to keep changes in musl minimal.